libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt

```diff
@@ -0,0 +1,296 @@
+ * /**
+ *  * Description of the RequiredExample. This is a data class because it is responsible for
+ *  * transporting data values.
+ *  *
+ *  * @param required description of what to expect from required
+ *  * @param foo description of what to expect from foo
+ *  * @param bar description of what to expect from bar
+ *  */
+ * class Example private constructor(
+ *     val required: String, // required
+ *     val foo: String, // optional with default value
+ *     val bar: Int? // optional without default value
+ * ) {
+ *     /**
+ *      * @return the builder that created the [Example]
+ *      */
+ *     fun toBuilder() = Builder(required).also {
+ *         it.foo(this.foo)
+ *         it.bar(this.bar)
+ *     }
+ *
+ *     override fun equals(other: Any?): Boolean {
+ *         if (this === other) return true
+ *         if (javaClass != other?.javaClass) return false
+ *
+ *         other as Example
+ *
+ *         if (required != other.required) return false
+ *         if (foo != other.foo) return false
+ *         if (bar != other.bar) return false
+ *
+ *         return true
+ *     }
+ *
+ *     override fun hashCode(): Int {
+ *         var result = required.hashCode()
+ *         result = 31 * result + foo.hashCode()
+ *         result = 31 * result + (bar ?: 0)
+ *         return result
+ *     }
+ *
+ *     override fun toString(): String {
+ *         return "Example(required='$required', foo='$foo', bar=$bar)"
+ *     }
+ *
+ *     /**
+ *      * Description for when to use this builder
+ *      *
+ *      * @param required does not have a default value
+ *      */
+ *     class Builder(
+ *         private val required: String
+ *     ) {
+ *         private var foo: String = "default foo value"
+ *         private var bar: Int? = null
+ *
+ *         /**
+ *          * @param foo description of what to expect when writing foo
+ *          * @return Builder
+ *          */
+ *         fun foo(foo: String) =
+ *             apply { this.foo = foo }
+ *
+ *         /**
+ *          * @param bar description of what to expect when writing bar
+ *          * @return Builder
+ *          */
+ *         fun bar(bar: Int?) =
+ *             apply { this.bar = bar }
+ *
+ *         /**
+ *          * Build new instance of [Example]
+ *          * @return Example
+ *          */
+ *         fun build(): Example {
+ *             return Example(
+ *                 required,
+ *                 foo,
+ *                 bar
+ *             )
+ *         }
+ *     }
+ * }
+ * ```
+ */
+abstract class BuilderTest<Implementation : Any, Builder> {
+
+    private val implClass: KClass<*> by lazy { getImplementationClass() }
+
+    /**
+     * Your builder's resulting type.
+     */
+    abstract fun getImplementationClass(): KClass<Implementation>
+
+    /**
+     * Make sure that all of the required and optional values are provided and that they distinct from default ones.
+     */
+    abstract fun getFilledUpBuilder(): Builder
```

Comment:
if we go with using inheritance, consider returning a list<Builder> here so we can test other scenarios. not sure about failure scenarios

Reply:
Currently, we are already iterating through all of the optional fields and testing equality for each one of them, so it's pretty exhaustive. I'm not sure how this could be automated when not all of the params are provided, we need all param instances anyway to perform the tests. I tried creating mocks using reflection, but because they are generic, I run into a lot of issues and didn't want to invest more time into it.